### PR TITLE
Re-implement `render` UI using terminal UI components (based on PTerm.sh)

### LINKED ISSUE #433: Re-implement render UI using terminal UI components (based on PTerm.sh)
## Approach

Should answer most of the following questions:

- Which docker compose files does render use? 

- Does it take into account any of my recent updates?

- What is it exactly that is rendered?

- How do I now deploy?

## Mockup

` .: ` denotes a progress bar.
` -> ` denotes a state change.

```shell
Render: render Kubernetes manifests in selected format.

» Validating compose sources...
Detecting secrets in: docker-compose.yaml 	
✓ Non detected in service: db
! Detected in service: wordpress
  | env var [WORDPRESS_DB_PASSWORD]
  | env var [WORDPRESS_DB_HOST]
  | env var [WORDPRESS_DB_USER]

Validation successful! 
However, to prevent secrets leaking, see help page: http://secrets-help-page 

» Validating deployment environments...
Detecting secrets in: docker-compose.dev.yaml 
✓ Non detected in service: db
✓ Non detected in service: cache
✓ Non detected in service: wordpress

Validation successful! 

» Detecting project updates...
Scanning dev: docker-compose.kev.dev.yaml
 ✓  No version update detected
 ✓  No service additions detected
 ✓  No service removals detected
 ✓  No env var removals detected
 ✓  No volume additions detected
 ✓  No volume removals detected
Scanning prod: docker-compose.kev.prod.yaml
 ✓  No version update detected
 ✓  No service additions detected
 ✓  No service removals detected
 ✓  No env var removals detected
 ✓  No volume additions detected
 ✓  No volume removals detected
Scanning stage: docker-compose.kev.stage.yaml
 ✓  No version update detected
 ✓  No service additions detected
 ✓  No service removals detected
 ✓  No env var removals detected
 ✓  No volume additions detected
 ✓  No volume removals detected

» Rendering manifests, format: kubernetes...
Rendering sandbox environment dev: docker-compose.kev.dev.yaml
.: Converting service: db   ->   ✓ Converted service: db
  | rendered StatefulSet: db-statefulset.yaml
  | rendered Service: db-service.yaml
.: Converting service: wordpress   ->   ✓ Converted service: wordpress
  | rendered Deployment: wordpress-deployment.yaml
  | rendered Service: db-service.yaml
.: Converting volume: db_data   ->   ✓ Converted volume: db_data
  | rendered PersistentVolumeClaim: db-data-persistentvolumeclaim.yaml
.: Creating networking policies   ->   ✓ Created networking policies 
  | rendered NetworkPolicy: default-networkpolicy.yaml
Rendering environment prod: docker-compose.kev.prod.yaml
.: Converting service: db   ->   ✓ Converted service: db
  | rendered StatefulSet: db-statefulset.yaml
  | rendered Service: db-service.yaml
.: Converting service: wordpress   ->   ✓ Converted service: wordpress
  | rendered Deployment: wordpress-deployment.yaml
  | rendered Service: db-service.yaml
.: Converting volume: db_data   ->   ✓ Converted volume: db_data
  | rendered PersistentVolumeClaim: db-data-persistentvolumeclaim.yaml
.: Creating networking policies   ->   ✓ Created networking policies 
  | rendered NetworkPolicy: default-networkpolicy.yaml

Project rendered!

A set of Kubernetes manifests has been created:
(Path to manifests based on <dir> setting)

The project can now be deployed to a Kubernetes cluster.

To test locally using the dev sandbox:
 - Ensure you have a local cluster up and running with a configured context.
 - Create a namespace: `kubectl create ns dev`.
 - Apply the manifests to the cluster: `kubectl apply -f <dir>/dev -n dev`.
 - Port forward to the main service: `kubectl port-forward service/<service_name> <service_port>:<destination_port>`

<kev only>
Check our tutorials for other examples of deploying your project, including CI:
https://github.com/appvia/kev/tree/master/docs/tutorials
</kev only>

```

### PR DESCRIPTION
**Pull request by ezodude**
Resolves #433 

**Pending**

- [x] Complete Rendering manifests UI.
- [x] Adding error UI sections.
- [x] Updating tests.

## Sample Output

```shell
$ kev render -e prod

» Loading...

» Validating compose sources...
Detecting secrets in: docker-compose.yaml
 !  Detected in service: db-test
   | env var [MYSQL_USER] - Contains word: user
   | env var [MYSQL_PASSWORD] - Contains word: password
   | env var [MYSQL_ROOT_PASSWORD] - Contains word: password
 !  Detected in service: wordpress
   | env var [WORDPRESS_DB_HOST] - Contains word: host
   | env var [WORDPRESS_DB_USER] - Contains word: user
   | env var [WORDPRESS_DB_PASSWORD] - Contains word: password

Validation successful!
However, to prevent secrets leaking, see help page:
https://github.com/appvia/kev/blob/master/docs/reference/config-params.md#reference-k8s-secret-key-value

» Verifying Skaffold...
 ✓  Skaffold manifest is available

» Validating deployment environments...
Detecting secrets in: docker-compose.kev.prod.yaml
 ✓  Non detected in service: db-test
 ✓  Non detected in service: wordpress

Validation successful!

» Detecting project updates...
prod: docker-compose.kev.prod.yaml
 ✓  No version update detected
 ✓  No service additions detected
 ✓  No service removals detected
 ✓  No env var removals detected
 ✓  No volume additions detected
 ✓  No volume removals detected

» Rendering manifests, format: kubernetes...
prod: docker-compose.kev.prod.yaml
 ✓  Converted service: db-test
   | rendered StatefulSet
   | rendered Service
   | rendered PersistentVolumeClaim
 ✓  Converted service: wordpress
   | rendered Deployment
   | rendered Service
 ✓  Created networking policies
   | rendered NetworkPolicy

Project manifests rendered!
A set of 'kubernetes' manifests have been generated:
  prod: k8s/prod

The project can now be deployed to a Kubernetes cluster.

To test locally:
 - Ensure you have a local cluster up and running with a configured context.
 - Create a namespace: `kubectl create ns ns-example`.
 - Apply the manifests to the cluster: `kubectl apply -f <manifests-dir>/<env> -n ns-example`.
 - Discover the main service: `kubectl get svc -n ns-example`.
 - Port forward to the main service: `kubectl port-forward service/<service_name> <service_port>:<destination_port> -n ns-example`.
```

### CODE REVIEW

File: cmd/kev/cmd/render.go [hunk 1]
```diff
@@ -40,8 +39,8 @@ var renderCmd = &cobra.Command{
 	Long:  renderLongDesc,
 	PreRunE: func(cmd *cobra.Command, args []string) error {
 		fns := []func(cmd *cobra.Command, args []string) error{
-			runReconcileCmd,
-			runDetectSecretsCmd,
+			// runReconcileCmd,
```
Comment (mangas) [inline cmd/kev/cmd/render.go:42]: remove comments?

File: cmd/kev/cmd/render.go [hunk 2]
```diff
@@ -90,7 +89,7 @@ func init() {
 }
 
 func runRenderCmd(cmd *cobra.Command, _ []string) error {
-	cmdName := "Render"
+	// cmdName := "Render"
```
Comment (mangas) [inline cmd/kev/cmd/render.go:92]: same here

File: cmd/kev/cmd/render.go [hunk 3]
```diff
@@ -103,19 +102,28 @@ func runRenderCmd(cmd *cobra.Command, _ []string) error {
...
-
-	log.DebugTitlef("Output format: %s", format)
-	if err := kev.Render(workingDir, format, singleFile, dir, envs, nil); err != nil {
-		return displayError(err)
-	}
-
-	os.Stdout.Write([]byte("\n"))
-
-	return nil
+	// displayCmdStarted(cmdName)
```
Comment (mangas) [inline cmd/kev/cmd/render.go:105]: is this section still needed?

Comment (ezodude) [reply cmd/kev/cmd/render.go:105]: Nope, its legacy - will come out in the wash 🙂

File: pkg/kev/converter/kubernetes/transform.go [hunk 4]
```diff
@@ -166,12 +185,24 @@ func (k *Kubernetes) Transform() ([]runtime.Object, error) {
...
+				renderedNetworkPolicy = np
 			}
 		}
 
 		allobjects = append(allobjects, objects...)
 	}
 
+	if renderedNetworkPolicy != nil {
+		stepNetworkPolicies := sg.Add("Creating networking policies")
+		stepNetworkPolicies.Success("Created networking policies")
```
Comment (marcinc) [inline pkg/kev/converter/kubernetes/transform.go:197]: It seems we use the same text for a step and then when we update the status to success. Would it make sense to just define it once? Or should we add a suffice to that message when Success / Warn / Error occurs?

File: pkg/kev/envs.go [hunk 5]
```diff
@@ -164,6 +163,14 @@ func (e *Environment) GetVolume(name string) (VolumeConfig, error) {
 	return e.override.getVolume(name)
 }
 
+// GetVolume retrieves a specific volume by name from the environment's override volumes.
```
Comment (marcinc) [inline pkg/kev/envs.go:166]: I guess this comment isn't for this method?

File: pkg/kev/init_errors.go [hunk 6]
```diff
@@ -54,12 +55,12 @@ deployment environments. Without them a project cannot be initialised.
 
 	initStepParsingComposeConfig: {
 		Error: "Invalid compose source(s)!",
-		ErrorDetails: `
-The provided compose source or sources are invalid. 'kev' requires valid 
-compose source files - without them a project cannot be initialised. 
+		ErrorDetails: fmt.Sprintf(`
+The provided compose source or sources are invalid. '%s' requires valid
```
Comment (marcinc) [inline pkg/kev/init_errors.go:59]: small rephrase suggestion `compose source or sources` -> `compose source(s)`

File: pkg/kev/override.go [hunk 7]
```diff
@@ -136,16 +137,213 @@ func (o *composeOverride) volumesLabelsExpandedFrom(other *composeOverride) Volu
 	return out
 }
 
-// diff detects changes between an override against another override.
-func (o *composeOverride) diff(other *composeOverride) changeset {
-	return newChangeset(other, o)
+// diffAndPatch detects and patches all changes between a destination override and the current override.
+// A change is either a create, update or delete event.
+// A change targets an override's version, services or volumes and it's properties will depend on the actual target.
```
Comment (marcinc) [inline pkg/kev/override.go:142]: it's -> its

File: pkg/kev/override.go [hunk 8]
```diff
@@ -136,16 +137,213 @@ func (o *composeOverride) volumesLabelsExpandedFrom(other *composeOverride) Volu
...
+// 		Type:   "update",                 //string
+// 		Index:  index,                    // interface{} in this case: int
+// 		Parent: "labels",                 // string
+// 		Target: config.LabelServiceType,  // string
+// 		Value:  srcSvc.GetLabels()[config.LabelServiceType], // interface{} in this case: string
+// }
+//
+// ENV VARS NOTE:
+// The changeset deals with the docker-compose `environment` attribute as a special case:
+// - Env vars in overrides override a project's docker-compose env vars.
```
Comment (marcinc) [inline pkg/kev/override.go:159]: Small tweak to better readability 
`Env vars in overrides override...` -> `Env vars specified in docker compose overrides modify...`

File: pkg/kev/override.go [hunk 9]
```diff
@@ -136,16 +137,213 @@ func (o *composeOverride) volumesLabelsExpandedFrom(other *composeOverride) Volu
...
+// - A changeset will ONLY REMOVE an env var if it is removed from a project's docker-compose env vars.
+// - A changeset will NOT update or create env vars in deployment environments.
```
Comment (marcinc) [inline pkg/kev/override.go:161]: `vars in deployment environments.` -> `vars in an environment specific docker compose override file.` ?

File: pkg/kev/override.go [hunk 10]
```diff
@@ -136,16 +137,213 @@ func (o *composeOverride) volumesLabelsExpandedFrom(other *composeOverride) Volu
...
+// - To create useful diffs a project's docker-compose env vars will be taken into account.
```
Comment (marcinc) [inline pkg/kev/override.go:162]: `a project's docker-compose` - are we talking about the base project docker compose file?

File: pkg/kev/render.go [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+	r.UI.Header("Detecting project updates...")
+
+	if _, err := r.manifest.ReconcileConfig(r.config.envs...); err != nil {
+		return err
+	}
+
+	return r.manifest.Environments.Write()
+}
+
+func (r *RenderRunner) RenderManifests() (map[string]string, error) {
```
Comment (marcinc) [inline pkg/kev/render.go:138]: Would be ideal to leave one line comments for exported methods.

Comment (marcinc) [reply pkg/kev/render.go:138]: We use word `manifest` in quite a few places when referring to Kev manifest, Skaffold manifest and to Kubernetes manifests. Would be great to rename this method to make it more explicit as to what manifests we're talking about.

It'd be beneficial to make that distinction and stick to nomenclature in other places for clarity.

File: pkg/kev/converter/kubernetes/converter.go [hunk 12]
```diff
@@ -113,3 +127,20 @@ func (c *K8s) Render(singleFile bool,
...
+func getSortedEnvs(projects map[string]*composego.Project) []string {
+	var out []string
+	for env, _ := range projects {
+		out = append(out, env)
+	}
+	sort.Strings(out)
+	return out
+}
+
+func getEnvFile(env string, files map[string][]string) string {
```
[the remark this answers is not in the document]
Comment (ezodude) [reply pkg/kev/converter/kubernetes/converter.go:140]: Removed - a simple variable does the job.

File: pkg/kev/converter/kubernetes/transform.go [hunk 13]
```diff
@@ -152,6 +161,16 @@ func (k *Kubernetes) Transform() ([]runtime.Object, error) {
 			return nil, err
 		}
 
+		stepSvc.Success(fmt.Sprintf("Converted service: %s", pSvc.Name))
+		for _, object := range objects {
+			k.UI.Output(
+				fmt.Sprintf("rendered %s", object.GetObjectKind().GroupVersionKind().Kind),
```
[the remark this answers is not in the document]
Comment (ezodude) [reply pkg/kev/converter/kubernetes/transform.go:177]: As discussed on slack. This attempts to work around the fact that we don’t know if the objects will get rendered as 1 file or separate files. This approach:

- Simplifies reporting.
- Provides a reference to what was produced by each service.

Will leave as is for now - debug messaging was not removed.

File: pkg/kev/envs.go (hunk 5 above)
Comment (ezodude) [reply pkg/kev/envs.go:166]: Fixed!

File: cmd/kev/cmd/render.go (hunk 1 above)
Comment (marcinc) [reply cmd/kev/cmd/render.go:42]: I guess these need removing?

File: pkg/kev/kev.go [hunk 14]
```diff
@@ -60,12 +61,36 @@ func InitProjectWithOptions(workingDir string, opts ...Options) error {
...
+
 // Reconcile reconciles changes with docker-compose sources against deployment environments.
 func Reconcile(workingDir string) (*Manifest, error) {
 	m, err := LoadManifest(workingDir)
 	if err != nil {
 		return nil, err
 	}
+
+	// TODO(es) Remove this after render runner is completed
+	m.UI = kmd.NoOpUI()
```
Comment (marcinc) [inline pkg/kev/kev.go:92]: Is this meant to be here?

File: pkg/kev/project.go [hunk 15]
```diff
@@ -53,6 +120,24 @@ func WithSkaffold(c bool) Options {
 	}
 }
 
+func WithManifestFormat(c string) Options {
```
Comment (marcinc) [inline pkg/kev/project.go:123]: It'd be great to have the exported functions documented. Here and other places in this PR. 🙏

**Review by marcinc — COMMENTED** [review]
a few more comments.

File: cmd/kev/cmd/render.go (hunk 1 above)
Comment (ezodude) [reply cmd/kev/cmd/render.go]: Removed

File: pkg/kev/project.go (hunk 15 above)
Comment (ezodude) [reply pkg/kev/project.go:123]: They're documented - I think I missed this one.

File: pkg/kev/kev.go (hunk 14 above)
Comment (ezodude) [reply pkg/kev/kev.go:92]: I'll update the TODO, but for now this is required while legacy `reconcile` cmd is still used in the `dev` cmd.

File: pkg/kev/project.go (hunk 15 above)
Comment (ezodude) [reply pkg/kev/project.go:127]: Done

**Review by marcinc — APPROVED** [review]
👍

**Pull request merged**